### ISSUE TITLE
Accessibility violation around overriding internal member is not detected by the compiler, causes a TypeLoadException exception at runtime

### ISSUE THREAD

**Issue by AlekseyTs** [dotnet/roslyn#33329]
Compile lib1.dll ```csc /t:library lib1.cs```:
```
[assembly: System.Runtime.CompilerServices.InternalsVisibleTo("lib2")]
[assembly: System.Runtime.CompilerServices.InternalsVisibleTo("lib3")]

public class A
{
    internal virtual void M() => System.Console.WriteLine("A");
}
```

Compile lib2.dll ```csc /t:library lib2.cs /r:lib1.dll```:
```
public class B : A
{
    internal override void M() => System.Console.WriteLine("B");
}
```

Compile lib3.exe ```csc lib3.cs /out:lib3.exe /r:lib1.dll /r:lib2.dll```
```
    public class C : B
    {
        internal override void M() => System.Console.WriteLine("C");

        static void Main()
        {
            new C().Test();
        }

        void Test()
        {
            base.M();
            M();
        }
    }
```

Run or PEVerify lib3.exe.
```
>lib3.exe

Unhandled Exception: System.TypeLoadException: Method 'M' on type 'C' from assembly 'lib3, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null' is overriding a method that is not
 visible from that assembly.

>peverify lib3.exe

Microsoft (R) .NET Framework PE Verifier.  Version  4.0.30319.0
Copyright (c) Microsoft Corporation.  All rights reserved.

[token  0x02000002] Type load failed.
1 Error(s) Verifying lib3.exe
```

Expected: 
Either execution should succeed and no exception should be thrown, the overriding works. Or, compiler should report an accessibility error for the overriding.

Now comment out the ```internal override void M() => System.Console.WriteLine("C");``` line in lib3.cs and rebuild lib3.exe.

Run or PEVerify lib3.exe:
```
>lib3.exe
A
B

>peverify lib3.exe

Microsoft (R) .NET Framework PE Verifier.  Version  4.0.30319.0
Copyright (c) Microsoft Corporation.  All rights reserved.

All Classes and Methods in lib3.exe Verified.
```

Observed: 
Note that ```base.M();``` prints "A", i.e. A.M is invoked instead of an inaccessible most derived override B.M. B.M is successfully executed by a virtual call ```M();```.

Expected:
The ```base.M();``` should report an error about the fact that B.M is not accessible.

**Comment by AlekseyTs** [comment]
Note that VB behavior is better ```vbc lib3.vb /out:lib3.exe /r:lib1.dll /r:lib2.dll```:
```
public class C 
    Inherits B

    Friend overrides Sub M()
        System.Console.WriteLine("C")
    End Sub

    Shared Sub Main()
        call new C().Test()
    End Sub

    Sub Test()
        MyBase.M()
        M()
    End Sub
End Class
```

Observed:
```
lib3.vb(4) : error BC30981: 'Friend Overrides Sub M()' in class 'C' cannot override 'Friend Overridable Overloads Sub M()' in class 'A' because an intermediate class 'B' overrides 'Friend Overridable Overloads Sub M()' in class 'A' but is not accessible.

    Friend overrides Sub M()
                         ~
```

However, if the illegal override is commented out, runtime behavior is consistent with C#. I am not sure if we can make this an error due to backward compatibility. But, it feels like the illegal override should be made an error in C#.

**Comment by gafter** [comment]
This appears to be a report of more than one issue.  The first issue would appear to be a runtime problem, and therefore more appropriately reported in `coreclr`.  The second issue may be a misunderstanding of what `base.M()` is supposed to do.  In this case it invokes the most derived accessible method `M` found from the base type.

**Comment by gafter** [comment]
We need to investigate if the first issue is a compiler bug or a runtime (verifier) bug.

The second appears to be by design.
